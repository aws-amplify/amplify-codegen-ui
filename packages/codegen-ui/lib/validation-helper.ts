/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import * as yup from 'yup';
import { InvalidInputError } from './errors';
import { StudioGenericEvent } from './types';

const alphaNumString = () => {
  return yup.string().matches(/^[a-zA-Z0-9]*$/, { message: 'Expected an alphanumeric string' });
};

const alphaNumNoLeadingNumberString = () => {
  return yup
    .string()
    .matches(/^[a-zA-Z][a-zA-Z0-9]*$/, { message: 'Expected an alphanumeric string, starting with a character' });
};

const propertiesSchema = (value: Object) => {
  return yup.object().shape(
    Object.fromEntries(
      Object.keys(value || {}).map((key) => [
        key,
        yup
          .object()
          .test('property', 'property cannot be empty.', (property: Object) => Object.keys(property).length > 0)
          .required(),
      ]),
    ),
  );
};

const eventsSchema = yup
  .object(
    Object.fromEntries(
      Object.keys(StudioGenericEvent)
        .filter((eventType) => Number.isNaN(Number(eventType)))
        .map((eventType) => [eventType, yup.object().nullable()]),
    ),
  )
  .noUnknown();

const schemaVersionSchema = () => {
  const versionNumberRegex = /^(\d+)\.(\d+)$/;
  const supportedMajorVersion = '1';
  const supportedMinorVersion = '0';
  return (
    yup
      .string()
      // eslint-disable-next-line no-template-curly-in-string
      .test('schemaVersion', 'unsupported schemaVersion ${originalValue}', (value: string | undefined) => {
        if (value === undefined) {
          return true;
        }
        const match = value.match(versionNumberRegex);
        if (match === null) {
          return false;
        }
        const [, majorVersion, minorVersion] = match;
        if (majorVersion !== supportedMajorVersion || minorVersion > supportedMinorVersion) {
          return false;
        }
        return true;
      })
  );
};

/**
 * Component Schema Definitions
 */
const studioComponentChildSchema: any = yup.object({
  componentType: alphaNumNoLeadingNumberString().required(),
  name: yup.string().required(),
  properties: yup.lazy((value) => propertiesSchema(value).required()),
  // Doing lazy eval here since we reference our own type otherwise
  children: yup.lazy(() => yup.array(studioComponentChildSchema.default(undefined))),
  figmaMetadata: yup.object().nullable(),
  variants: yup.array().nullable(),
  overrides: yup.object().nullable(),
  bindingProperties: yup.object().nullable(),
  collectionProperties: yup.object().nullable(),
  actions: yup.object().nullable(),
  events: eventsSchema.nullable(),
});

const studioComponentSchema = yup
  .object({
    name: alphaNumString().required(),
    id: yup.string().nullable(),
    sourceId: yup.string().nullable(),
    schemaVersion: yup.lazy(() => schemaVersionSchema().nullable()),
    componentType: alphaNumNoLeadingNumberString().required(),
    properties: yup.lazy((value) => propertiesSchema(value).required()),
    children: yup.array(studioComponentChildSchema).nullable(),
    figmaMetadata: yup.object().nullable(),
    variants: yup.array().nullable(),
    overrides: yup.object().nullable(),
    bindingProperties: yup.object().nullable(),
    collectionProperties: yup.object().nullable(),
    actions: yup.object().nullable(),
    events: eventsSchema.nullable(),
  })
  // eslint-disable-next-line func-names
  .test('unique-component-names', 'All component names must be unique', function (value) {
    const { path, createError } = this;
    const { duplicateNames } = findCollidingNames(value);
    if (duplicateNames.size > 0) {
      return createError({
        path,
        message: `Duplicate names are not allowed within a component, found duplicates of ${JSON.stringify([
          ...duplicateNames,
        ])}`,
      });
    }
    return true;
  });

type ComponentNameMetadata = {
  names: Set<string>;
  duplicateNames: Set<string>;
};

const findCollidingNames = (value: any): ComponentNameMetadata => {
  const names = value.name ? new Set([value.name]) : new Set();
  const componentNameMetadata: ComponentNameMetadata = { names, duplicateNames: new Set() };
  if (!value.children) {
    return componentNameMetadata;
  }
  const childComponentNameMetadata = value.children
    .map((child: any) => findCollidingNames(child))
    .reduce(
      (previous: ComponentNameMetadata, next: ComponentNameMetadata) => mergeComponentNameMetadata(previous, next),
      { names: new Set(), duplicateNames: new Set() },
    );
  return mergeComponentNameMetadata(componentNameMetadata, childComponentNameMetadata);
};

const mergeComponentNameMetadata = (lhs: ComponentNameMetadata, rhs: ComponentNameMetadata): ComponentNameMetadata => {
  const names = new Set([...lhs.names, ...rhs.names]);
  const duplicateNames = new Set([...lhs.duplicateNames, ...rhs.duplicateNames]);
  lhs.names.forEach((lhsName) => {
    if (rhs.names.has(lhsName)) {
      duplicateNames.add(lhsName);
    }
  });
  return { names, duplicateNames };
};

/**
 * Theme Schema Definitions
 */
const studioThemeValuesSchema: any = yup.object({
  key: yup.string().required(),
  value: yup
    .object({
      value: yup.string(),
      children: yup.lazy(() => yup.array(studioThemeValuesSchema.default(undefined))),
    })
    .required(),
});

const studioThemeSchema = yup.object({
  name: alphaNumString().required(),
  id: yup.string().nullable(),
  values: yup.array(studioThemeValuesSchema).required(),
  overrides: yup.array(studioThemeValuesSchema).nullable(),
});

/**
 * Form Schema Definitions
 */
const studioFormSchema = yup.object({
  name: alphaNumString().required(),
  id: yup.string().nullable(),
  formActionType: yup.string().matches(new RegExp('(create|update)')),
  dataType: yup.object({
    dataSourceType: yup.string().matches(new RegExp('(DataStore|Custom)')),
    dataTypeName: yup.string().required(),
  }),
  fields: yup.object().nullable(),
  sectionalElements: yup.object().nullable(),
  style: yup.object().nullable(),
});

/**
 * Studio Schema Validation Functions and Helpers.
 */
const validateSchema = (validator: yup.AnySchema, studioSchema: any) => {
  try {
    validator.validateSync(studioSchema, { strict: true, abortEarly: false });
  } catch (e) {
    if (e instanceof yup.ValidationError) {
      throw new InvalidInputError(e.errors.join(', '));
    }
    throw e;
  }
};

export const validateComponentSchema = (schema: any) => validateSchema(studioComponentSchema, schema);
export const validateThemeSchema = (schema: any) => validateSchema(studioThemeSchema, schema);
export const validateFormSchema = (schema: any) => validateSchema(studioFormSchema, schema);
