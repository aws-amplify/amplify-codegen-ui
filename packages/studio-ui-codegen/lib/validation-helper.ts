/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import * as yup from 'yup';
import { InvalidInputError } from './errors';

const alphaNumString = () => {
  return yup.string().matches(/^[a-zA-Z0-9]*$/, { message: 'Expected an alphanumeric string' });
};

const alphaNumNoLeadingNumberString = () => {
  return yup
    .string()
    .matches(/^[a-zA-Z][a-zA-Z0-9]*$/, { message: 'Expected an alphanumeric string, starting with a character' });
};

/**
 * Component Schema Definitions
 */
const studioComponentChildSchema: any = yup.object({
  componentType: alphaNumNoLeadingNumberString().required(),
  // TODO: Name is required in the studio-types file, but doesn't seem to need to be. Relaxing the restriction here.
  name: yup.string(),
  properties: yup.object().required(),
  // Doing lazy eval here since we reference our own type otherwise
  children: yup.lazy(() => yup.array(studioComponentChildSchema.default(undefined))),
  figmaMetadata: yup.object(),
  variants: yup.array(),
  overrides: yup.object(),
  bindingProperties: yup.object(),
  collectionProperties: yup.object(),
  actions: yup.object(),
});

const studioComponentSchema = yup.object({
  name: alphaNumString(),
  id: yup.string(),
  sourceId: yup.string(),
  componentType: alphaNumNoLeadingNumberString().required(),
  properties: yup.object().required(),
  children: yup.array(studioComponentChildSchema),
  figmaMetadata: yup.object(),
  variants: yup.array(),
  overrides: yup.object(),
  bindingProperties: yup.object(),
  collectionProperties: yup.object(),
  actions: yup.object(),
});

/**
 * Theme Schema Definitions
 */
const studioThemeValuesSchema: any = yup.object({
  key: yup.string().required(),
  value: yup
    .object({
      value: yup.string(),
      children: yup.lazy(() => yup.array(studioThemeValuesSchema.default(undefined))),
    })
    .required(),
});

const studioThemeSchema = yup.object({
  name: alphaNumString().required(),
  id: yup.string(),
  values: yup.array(studioThemeValuesSchema).required(),
  overrides: yup.array(studioThemeValuesSchema),
});

/**
 * Studio Schema Validation Functions and Helpers.
 */
const validateSchema = (validator: yup.AnySchema, studioSchema: any) => {
  try {
    validator.validateSync(studioSchema, { strict: true, abortEarly: false });
  } catch (e) {
    if (e instanceof yup.ValidationError) {
      throw new InvalidInputError(e.errors.join(', '));
    }
    throw e;
  }
};

export const validateComponentSchema = (schema: any) => validateSchema(studioComponentSchema, schema);
export const validateThemeSchema = (schema: any) => validateSchema(studioThemeSchema, schema);
