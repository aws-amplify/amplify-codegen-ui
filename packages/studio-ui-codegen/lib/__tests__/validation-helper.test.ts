/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { validateComponentSchema, validateThemeSchema } from '../validation-helper';

describe('validation-helper', () => {
  describe('validateComponentSchema', () => {
    test('throws no error on valid type', () => {
      validateComponentSchema({
        name: 'MyBindingView',
        componentType: 'View',
        properties: {},
      });
    });

    test('top-level component requires componentType', () => {
      expect(() => {
        validateComponentSchema({
          name: 'MyBindingView',
          properties: {},
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('top-level component requires properties', () => {
      expect(() => {
        validateComponentSchema({
          componentType: 'View',
          name: 'MyBindingView',
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('top-level component requires componentType to be the correct type', () => {
      expect(() => {
        validateComponentSchema({
          componentType: 2,
          name: 'MyBindingView',
          properties: {},
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('child component requires componentType to be the correct type', () => {
      expect(() => {
        validateComponentSchema({
          componentType: 'View',
          name: 'MyBindingView',
          properties: {},
          children: [
            {
              componentType: 3,
              properties: {},
            },
          ],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('deeply nested child components requires properties', () => {
      expect(() => {
        validateComponentSchema({
          componentType: 'View',
          name: 'MyBindingView',
          properties: {},
          children: [
            {
              componentType: 'View',
              properties: {},
              children: [
                {
                  componentType: 'View',
                  properties: {},
                  children: [
                    {
                      componentType: 'Button',
                    },
                  ],
                },
              ],
            },
          ],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('fails on componentType with whitespace', () => {
      expect(() => {
        validateComponentSchema({
          name: 'CustomComponent',
          componentType: 'View 2',
          properties: {},
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('fails on componentType with leading number', () => {
      expect(() => {
        validateComponentSchema({
          name: 'CustomComponent',
          componentType: '2View',
          properties: {},
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('succeeds on componentType with trailing number', () => {
      validateComponentSchema({
        name: 'CustomComponent',
        componentType: 'View2',
        properties: {},
      });
    });

    test('child component name may contain whitespace', () => {
      validateComponentSchema({
        name: 'CustomComponent',
        componentType: 'View',
        properties: {},
        children: [
          {
            name: 'I Have Spaces',
            componentType: 'Button',
            properties: {},
          },
        ],
      });
    });
  });

  describe('validateThemeSchema', () => {
    test('throws no error on valid type', () => {
      validateThemeSchema({
        name: 'MyBindingView',
        values: [
          {
            key: 'breakpoints',
            value: {},
          },
          {
            key: 'tokens',
            value: {},
          },
        ],
        overrides: [],
      });
    });

    test('top-level component requires name', () => {
      expect(() => {
        validateThemeSchema({
          values: [
            {
              key: 'breakpoints',
              value: {},
            },
            {
              key: 'tokens',
              value: {},
            },
          ],
          overrides: [],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('top-level component requires values', () => {
      expect(() => {
        validateThemeSchema({
          name: 'MyBindingView',
          overrides: [],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('children objects should not be empty', () => {
      expect(() => {
        validateThemeSchema({
          name: 'MyBindingView',
          values: [
            {
              key: 'breakpoints',
              value: {},
            },
            {},
          ],
          overrides: [],
        });
      }).toThrowErrorMatchingSnapshot();
    });

    test('overrides should contain the right shape', () => {
      expect(() => {
        validateThemeSchema({
          name: 'MyBindingView',
          values: [],
          overrides: [
            {
              value: {},
            },
          ],
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
