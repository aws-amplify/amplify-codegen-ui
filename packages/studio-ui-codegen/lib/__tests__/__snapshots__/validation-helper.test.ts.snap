// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation-helper validateComponentSchema child component requires componentType to be the correct type 1`] = `"children[0].componentType must be a \`string\` type, but the final value was: \`3\`."`;

exports[`validation-helper validateComponentSchema deeply nested child components requires properties 1`] = `"children[0].children[0].children[0].properties is a required field"`;

exports[`validation-helper validateComponentSchema fails on componentType with leading number 1`] = `"Expected an alphanumeric string, starting with a character"`;

exports[`validation-helper validateComponentSchema fails on componentType with whitespace 1`] = `"Expected an alphanumeric string, starting with a character"`;

exports[`validation-helper validateComponentSchema top-level component requires componentType 1`] = `"componentType is a required field"`;

exports[`validation-helper validateComponentSchema top-level component requires componentType to be the correct type 1`] = `"componentType must be a \`string\` type, but the final value was: \`2\`."`;

exports[`validation-helper validateComponentSchema top-level component requires properties 1`] = `"properties is a required field"`;

exports[`validation-helper validateThemeSchema children objects should not be empty 1`] = `"values[1].key is a required field, values[1].value is a required field"`;

exports[`validation-helper validateThemeSchema overrides should contain the right shape 1`] = `"overrides[0].key is a required field"`;

exports[`validation-helper validateThemeSchema top-level component requires name 1`] = `"name is a required field"`;

exports[`validation-helper validateThemeSchema top-level component requires values 1`] = `"values is a required field"`;
